src/lender/coordinator.sol

```diff
@@ -125,97 +142,264 @@ contract EpochCoordinator is Ticker, Auth  {
+            return true;
+        }
+        return false;
+    }
+
+    function _improvementScoreCase(uint seniorRedeem, uint juniorRedeem,
+        uint juniorSupply, uint seniorSupply) internal returns(int) {
+        Fixed27 memory currSeniorRatio = Fixed27(calcSeniorRatio(safeAdd(epochSeniorBalance, epochSeniorDebt),
+            epochNAV, epochReserve));
+
+        if(bestSubScore == 0) {
+            // define no orders score as benchmark if no previous submission exists
+            bestSubScore = scoreImprovement(currSeniorRatio, currSeniorRatio, epochReserve);
+        }
+
+        uint newReserve = calcNewReserve(seniorRedeem, juniorRedeem, seniorSupply, juniorSupply);
 
-        // solution is new best => 0
+        Fixed27 memory newSeniorRatio = Fixed27(calcSeniorRatio(calcNewSeniorAssetValue( seniorRedeem, seniorSupply,
+            epochSeniorDebt, epochSeniorBalance, newReserve), epochNAV, newReserve));
+
+        uint score =  scoreImprovement(newSeniorRatio, currSeniorRatio, newReserve);
+
+        if (score < bestSubScore) {
+            // solution is not the best => -1
+            return -1;
+        }
+
+        // solution doesn't satisfy all pool constraints but improves the current violation
+        saveNewOptimum(seniorRedeem, juniorRedeem, juniorSupply, seniorSupply, score);
         return 0;
     }
 
+    // returns the normalized distance (maxReserve = ONE) from the newReserve and maxReserve/2
+    function scoreDistanceReserve(Fixed27 memory newReserve_) public view returns (uint score) {
+        return rmul(1000, rdiv(ONE, abs(safeDiv(ONE, 2), newReserve_.value)));
+    }
+
+    function scoreImprovement(Fixed27 memory newSeniorRatio, Fixed27 memory currSeniorRatio,
+        uint newReserve_) public  returns(uint) {
+        (Fixed27 memory minSeniorRatio, Fixed27 memory maxSeniorRatio) = assessor.seniorRatioBounds();
+
+        // normalize reserve by defining maxReserve as ONE
+        Fixed27 memory normalizedNewReserve = Fixed27(rdiv(newReserve_, assessor.maxReserve()));
+
+        // current ratio is healthy
+        if (checkRatioInRange(currSeniorRatio, minSeniorRatio, maxSeniorRatio) == true) {
+
+            // the new proposed solution would violate the ratio constraints
+            if (checkRatioInRange(newSeniorRatio, minSeniorRatio, maxSeniorRatio) == false)
+            {
+                return 0;
+            }
+            // only points for maxRatio improvement
+
+            return scoreDistanceReserve(normalizedNewReserve);
+
+        }
+
+        // gas optimized implementation
+        // abs of ratio can never be zero
+        uint score = rmul(10000, rdiv(ONE, abs(newSeniorRatio.value,
+            safeDiv(safeAdd(minSeniorRatio.value, maxSeniorRatio.value), 2))));
+
+        // ratio constraints and maxReserve are in the current state violated
+        // additional score
+        if (epochReserve >= assessor.maxReserve()) {
```

Comment:
since fixing the ratio might actually put the reserve in a worse state then before, shouldn't we always compute the distanceScore of the reserve

Reply:
Same for this issue. I will write it in the agenda for the meeting